### PR TITLE
in loadMany, return resulting deferred object to allow further manipulat...

### PR DESCRIPTION
**Pull request by bebbi**
...ion

### CODE REVIEW

**Comment by concavelenz** [comment]
Needs CLA: https://github.com/google/closure-library/blob/master/CONTRIBUTING

**Comment by bebbi** [comment]
cla done

**Comment by bebbi** [comment]
@concavelenz  why label cla:no? CLA is "signed and filed" March 10 (via email confirmation from google)

**Comment by bebbi** [comment]
fixed author now matching signed CLA (re-write commit)

**Comment by bebbi** [comment]
verified email in github too in case that's necessary

File: closure/goog/net/jsloader.js [hunk 1]
```diff
@@ -119,8 +119,9 @@ goog.net.jsloader.loadMany = function(uris, opt_options) {
     if (uris.length) {
       deferred.addBoth(popAndLoadNextScript);
     }
+    return deferred;
   };
-  popAndLoadNextScript();
+  return popAndLoadNextScript();
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: Add a @return type to loadMany.

Comment (bebbi) [inline closure/goog/net/jsloader.js]: ok will do in a week or so.

**Comment by joeltine** [comment]
Updates on this?

**Comment by bebbi** [comment]
done. (rebased commit, annotated return and simplified code.) Synopsis for PR:

Fixes issue where caller doesn't know when load sequence is finished in loadMany
Fixes issue where it's impossible to parallelize script loading with multiple independent loadMany calls, e.g.: loadMany([A, B, C]) and loadMany([D, E, F])
Impact: since caller knows what's going on, there is no need for a goog.net.jsloader.scriptsToLoad_ queue. Clients can avoid the race condition mentioned.

File: closure/goog/net/jsloader.js [hunk 2]
```diff
@@ -94,36 +94,26 @@ goog.net.jsloader.scriptsToLoad_ = [];
...
-  // sequence for client B, even though client A's scripts haven't all been
-  // loaded yet.
-  //
-  // To work around this issue, all module loads share a queue.
+  // The client receives a deferred to manage parallel loads.
   if (!uris.length) {
-    return;
-  }
-
-  var isAnotherModuleLoading = goog.net.jsloader.scriptsToLoad_.length;
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: Ehhh, I think I'd prefer you not change the behavior of this method this radically. The original impl was designed with a specific use case in mind and breaking that assumption could cause breakages internally.

Can you revert this back to the original fix w/ only `@return` added?

Comment (bebbi) [inline closure/goog/net/jsloader.js]: Ok. I had asked for feedback on the google group. It's a very specific use-case. The problem I see is it blocks other use-cases, as the caller has no control. Do you agree w/ some option to use/skip the queue? A param (opt_queue), or maybe an add'l func? Tx

Comment (joeltine) [inline closure/goog/net/jsloader.js]: I think the addition is reasonable. I wouldn't make it the default though. Adding another field to `opt_options` is probably the best way to do it.  Take a look at the `goog.net.jsloader.Options` type.

That said, can you do that in a separate PR? This returning deferred change seems valuable enough on its own.

File: closure/goog/net/jsloader.js [hunk 3]
```diff
@@ -94,36 +94,26 @@ goog.net.jsloader.scriptsToLoad_ = [];
...
-  goog.array.extend(goog.net.jsloader.scriptsToLoad_, uris);
-  if (isAnotherModuleLoading) {
-    // jsloader is still loading some other scripts.
-    // In order to prevent the race condition noted above, we just add
-    // these URIs to the end of the scripts' queue and return.
-    return;
```
Comment (bebbi) [inline closure/goog/net/jsloader.js]: I’ll do as you suggest. However, a quick request for comment:
It seems to me that the code loses a tad elegance.
To support the original use-case, we now need to store active deferred to return it in the if (isAnotherModuleLoading) clause.
But since we always return deferreds now, that use-case could easily be handled by the caller.
I'm happy to code for it in loadMany, but as "with queue" is more specific than w/o (judged by the number of lines), it would seem more natural to use the option to enable it ("..plus one uri queue please") rather than disable it ("..oh, without that queue thing please").
I am aware that backwards compatibility can be a issue somewhere.
With that, could you confirm that you prefer it this way and I'll PR per your spec.

**Comment by bebbi** [comment]
ok nevrmind, going to submit per spec. Just let me know in case you change decision

**Comment by bebbi** [comment]
I'm seeing a linter issue for long line. Going to truncate all lines to 80 and re-submit

**Comment by bebbi** [comment]
See #523 for the PR with skipQueue

File: closure/goog/net/jsloader.js (hunk 3 above)
Comment (joeltine) [inline closure/goog/net/jsloader.js]: >But since we always return deferreds now, that use-case could easily be handled by the caller.

I don't know if I agree with this. Yes, it's possible for the caller to manage it, but it would require passing around a deferred to anything that wants to chain loads. The way it is now, your only dependency is on `jsloader`, and invoking `loadMany` means you are always guaranteed to get the latest deferred regardless of where/how you call it. IMO, this seems nicer.

Re: backwards compatibility, this is the cost of maintaining a widely used library. You have to weigh the cost and potential payoff of a change. If we make this change and I break 200 tests, is it really worth me spending 3 days fixing them for this change? Especially when it's just as easy to just add an option to enable the functionality you want without breaking anyone.

File: closure/goog/net/jsloader.js [hunk 4]
```diff
@@ -55,8 +55,8 @@ goog.net.jsloader.DEFAULT_TIMEOUT = 5000;
  * document: The HTML document under which to load the JavaScript. Default is
  *     the current document.
  * cleanupWhenDone: If true clean up the script tag after script completes to
- *     load. This is important if you just want to read data from the JavaScript
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: I'm seeing this as 80 cols. Were you getting lint errors?

File: closure/goog/net/jsloader.js [hunk 5]
```diff
@@ -122,15 +134,17 @@ goog.net.jsloader.loadMany = function(uris, opt_options) {
...
+  goog.net.jsloader.scriptLoadingDeferred_ = popAndLoadNextScript();
+  return goog.net.jsloader.scriptLoadingDeferred_;
 };
 
 
 /**
  * Loads and evaluates a JavaScript file.
  * When the script loads, a user callback is called.
- * It is the client's responsibility to verify that the script ran successfully.
+ * It's the client's responsibility to verify that the script ran successfully
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: End w/ period.

File: closure/goog/net/jsloader.js [hunk 6]
```diff
@@ -103,7 +113,7 @@ goog.net.jsloader.loadMany = function(uris, opt_options) {
   //
   // To work around this issue, all module loads share a queue.
   if (!uris.length) {
-    return;
+    return goog.async.Deferred.succeed(null);
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: passing null isn't necessary.

Comment (joeltine) [inline closure/goog/net/jsloader.js]: Actually, I'm guessing you're trying to emulate what `load` does? If so, that's fine.

File: closure/goog/net/jsloader.js [hunk 7]
```diff
@@ -122,15 +134,17 @@ goog.net.jsloader.loadMany = function(uris, opt_options) {
     if (uris.length) {
       deferred.addBoth(popAndLoadNextScript);
     }
+    return deferred;
   };
-  popAndLoadNextScript();
+  goog.net.jsloader.scriptLoadingDeferred_ = popAndLoadNextScript();
+  return goog.net.jsloader.scriptLoadingDeferred_;
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: Can you add a unit test that validates the deferred corresponds to the completion of loading all scripts?

**Comment by bebbi** [comment]
done. Comment changes reset, my bad

File: closure/goog/net/jsloader_test.js [hunk 8]
```diff
@@ -134,7 +138,24 @@ function testLoadMany() {
...
   };
+
+  result.addCallback(function() {
+    testCase.continueTesting();
+
+    // verify that the last loaded script callback has executed
+    assertEquals('verification object', 'Test #4 loaded', window.test4);
+  });
+
+
```
Comment (joeltine) [inline closure/goog/net/jsloader_test.js]: kill newlines

Comment (bebbi) [inline closure/goog/net/jsloader_test.js]: which ones? I've copied the writing style in that file, they're all over the place

Comment (joeltine) [inline closure/goog/net/jsloader_test.js]: Between function definitions, yes. Not at the end of a function block though.

**Comment by joeltine** [comment]
Good job on the test! I'm impressed. It's testing exactly what I expect and written in accordance with the surrounding style. Well done.

File: closure/goog/net/jsloader_test.js (hunk 8 above)
Comment (bebbi) [inline closure/goog/net/jsloader_test.js]: ah - just saw it now. tx

**Comment by bebbi** [comment]
hey thanks. hope we've got it now (for today at least that's it for me..). tx for patience. I've rebased #523 on top of this one, will add test once agreed.

File: closure/goog/net/jsloader.js [hunk 9]
```diff
@@ -78,6 +78,16 @@ goog.net.jsloader.scriptsToLoad_ = [];
 
 
 /**
+ * The deferred that is loading the uris in scriptsToLoad_
+ * We need to return this to a caller that wants to load uris while
+ * a deferred is already working on them
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: end w/ period.

File: closure/goog/net/jsloader.js [hunk 10]
```diff
@@ -111,8 +123,10 @@ goog.net.jsloader.loadMany = function(uris, opt_options) {
   if (isAnotherModuleLoading) {
     // jsloader is still loading some other scripts.
     // In order to prevent the race condition noted above, we just add
-    // these URIs to the end of the scripts' queue and return.
-    return;
+    // these URIs to the end of the scripts' queue and return the deferred
```
Comment (joeltine) [inline closure/goog/net/jsloader.js]: Please use proper capitalization and end sentence w/ period.

Comment (bebbi) [inline closure/goog/net/jsloader.js]: done. not sure what capitalization refers to. deferred? have changed sentence to work around.

Comment (joeltine) [inline closure/goog/net/jsloader.js]: Oh, I'm sorry. I'm talking about English grammar. https://en.wikipedia.org/wiki/Capitalization

Comment (bebbi) [inline closure/goog/net/jsloader.js]: I got that. But aside from the word 'deferred' (I tend to write instances in lower case but probably one can argue for upper case here), I didn't spot any case issues in other words that I added. Is it okay as amended?

Comment (joeltine) [inline closure/goog/net/jsloader.js]: OHHH. Haha, I thought "these" started a new sentence. A side effect of using the "unified" view. Sorry for confusion.

**Comment by joeltine** [comment]
Thanks!

**Comment by bebbi** [comment]
Thanks to you. I hope we can add #523 too. I've checked all grammer and linting already :)
